src/core/algorithms/dd/split/split.cpp

```diff
@@ -298,95 +352,80 @@ double Split::CalculateDistance(model::ColumnIndex column_index,
-        std::deque<model::PLI::Cluster> const& index = pli->GetIndex();
-        std::shared_ptr<std::vector<int> const> probing_table = pli->CalculateAndGetProbingTable();
-        model::PLI::Cluster const& pt = *probing_table.get();
+        DistancePositionListIndex pli(typed_relation_->GetColumnData(column_index), num_rows_);
+        std::vector<ClusterInfo> const& clusters = pli.GetClusters();
+        std::size_t const num_clusters = clusters.size();
+        std::vector<std::vector<double>> cur_column_distances;
+        cur_column_distances.reserve(num_clusters);
 
         double max_dif = 0, min_dif = std::numeric_limits<double>::max();
-        for (std::size_t i = 0; i < index.size(); i++) {
-            for (std::size_t j = 0; j < index.size(); j++) {
-                std::size_t const first_index = index[i][0];
-                std::size_t const second_index = index[j][0];
-                if (first_index < num_rows_ && second_index < num_rows_) {
-                    InsertDistance(column_index, first_index, second_index, min_dif, max_dif);
-                }
-            }
-        }
-        for (std::size_t i = 0; i < num_rows_; i++) {
-            for (std::size_t j = 0; j < num_rows_; j++) {
-                if (pt[i] != 0 && pt[j] != 0) {
-                    distances_[column_index][i][j] =
-                            distances_[column_index][index[pt[i] - 1][0]][index[pt[j] - 1][0]];
-                    if (i != j && pt[i] == pt[j]) {
-                        min_dif = 0;
-                    }
-                } else {
-                    InsertDistance(column_index, i, j, min_dif, max_dif);
-                }
+        for (ClusterIndex i = 0; i < num_clusters; i++) {
+            cur_column_distances.emplace_back();
+            cur_column_distances[i].reserve(num_clusters - i);
+            cur_column_distances[i].push_back(0);
+            for (ClusterIndex j = i + 1; j < num_clusters; j++) {
+                std::size_t const first_index = clusters[i].first_tuple_index;
+                std::size_t const second_index = clusters[j].first_tuple_index;
+                double const dif = CalculateDistance(column_index, {first_index, second_index});
+                max_dif = std::max(max_dif, dif);
+                min_dif = std::min(min_dif, dif);
+                cur_column_distances[i].push_back(dif);
             }
+            if (clusters[i].size > 1) min_dif = 0;
         }
         min_max_dif_[column_index] = {min_dif, max_dif};
+        distances_.emplace_back(std::move(cur_column_distances));
+        plis_[column_index] = std::move(pli);
     }
 }
 
 bool Split::IsFeasible(DF const& d) {
-    for (std::size_t i = 0; i < num_rows_; i++) {
-        for (std::size_t j = i + 1; j < num_rows_; j++) {
-            if (CheckDF(d, {i, j})) return true;
-        }
+    for (auto const& pair : tuple_pairs_) {
+        if (CheckDF(d, pair)) return true;
     }
     return false;
```

Comment:
```suggestion
    return std::ranges::any_of(tuple_pairs_, [this, &d](const auto& pair) { return CheckDF(d, pair); });
```

Reply:
maybe other methods could also be simplified in this way by using exact algorithm with readable name instead of a raw loop